### PR TITLE
feat (processor/k8sattributes): wait for synced when starting k8sattributes processor.

### LINKED ISSUE #32556: Some data couldn't be associated with k8s metadata when the agent was just started.
### Component(s)

processor/k8sattributes

## Description

When the agent is first started, the k8s metadata has not been fully synchronized by the k8sattributes processor. If any data passes through the component at this point, they will not be associated with metadata.

## Steps to Reproduce

- start a service that continuously sends a large amount of data to the agent.
- (re)start the agent.

## Expected Result

All data should be associated with k8s metadata.

## Actual Result

There may be a little data that hasn't be assosicated.

### Collector version

v0.95.0

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

I think it's the best solution to wait for an initial sync when starting k8sattributes processor.

Or we can block data util an initial sync is finished.

### PR DESCRIPTION
**Pull request by h0cheung**
**Description:** <Describe what has changed.>

When starting `k8sattributes` processor, block until an initial synchronization has been completed. This solves #32556 

**Link to tracking Issue:** <Issue number if applicable>

fix #32556

**Testing:** <Describe what testing was performed and which tests were added.>

Tested in a cluster with constant high span traffic, no more spans with unassociated k8s metadata after adding this pr.

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: processor/k8sattributesprocessor/processor.go [hunk 1]
```diff
@@ -71,7 +71,11 @@ func (kp *kubernetesprocessor) Start(_ context.Context, _ component.Host) error
 		}
 	}
 	if !kp.passthroughMode {
-		go kp.kc.Start()
+		err := kp.kc.Start()
+		if err != nil {
+			kp.telemetrySettings.ReportStatus(component.NewFatalErrorEvent(err))
+			return nil
```
Comment (TylerHelmuth) [inline processor/k8sattributesprocessor/processor.go:77]: We should return the error

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 2]
```diff
@@ -179,50 +180,61 @@ func New(logger *zap.Logger, apiCfg k8sconfig.APIConfig, rules ExtractionRules,
...
 		})
 		if err != nil {
-			c.logger.Error("error adding event handler to node informer", zap.Error(err))
+			return err
 		}
+		synced = append(synced, reg.HasSynced)
 		go c.nodeInformer.Run(c.stopCh)
 	}
+
+	if !cache.WaitForCacheSync(c.stopCh, synced...) {
```
Comment (TylerHelmuth) [inline processor/k8sattributesprocessor/internal/kube/client.go:233]: The godoc comments for this function recommend using `WaitForNamedCacheSync` instead.

Comment (TylerHelmuth) [reply processor/k8sattributesprocessor/internal/kube/client.go:233]: Also, does this function block indefinitely? That would be unacceptable, we'd need to be able to set a timeout for how long we wait to sync.

Comment (h0cheung) [reply processor/k8sattributesprocessor/internal/kube/client.go:233]: > The godoc comments for this function recommend using `WaitForNamedCacheSync` instead.

It seems that `WaitForNamedCacheSync` only wraps `WaitForCacheSync` with a logger, which prints logs in different format from our logger. Maybe it's better to handle logging by ourselves?

Comment (h0cheung) [reply processor/k8sattributesprocessor/internal/kube/client.go:233]: > Also, does this function block indefinitely? That would be unacceptable, we'd need to be able to set a timeout for how long we wait to sync.

Yeah. But I'm not sure whether it should be considered as a success with a warning, or a failure.

Comment (fatsheep9146) [reply processor/k8sattributesprocessor/internal/kube/client.go:233]: I saw the `WaitForCacheSync` called `wait.PollImmediateUntil` function, and the comments of wait.PollImmediateUntil is
```
// PollImmediateUntil tries a condition func until it returns true, an error or stopCh is closed.
```

So in my opinion, if some extreme case happened, this function will block for a long time.

How about using context.WithTimeout to create a context, the send to WaitForCacheSync in case the processor is blocked for a long time?

**Comment by orloffv** [comment]
@h0cheung Hi there, I was wondering if you have the time to finalize and merge the pull request. Looking forward to your update. Thanks!

File: .chloggen/k8sattributes-block.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: processor/k8sattributes
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Block when starting util the metadata have been synced, to fix that some data couldn't be associated with metadata when the agent was just started.
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: []
```
Comment (fatsheep9146) [inline .chloggen/k8sattributes-block.yaml:13]: please add 32556 here to pass the changlog check.

File: processor/k8sattributesprocessor/config.go [hunk 4]
```diff
@@ -38,6 +39,13 @@ type Config struct {
 	// Exclude section allows to define names of pod that should be
 	// ignored while tagging.
 	Exclude ExcludeConfig `mapstructure:"exclude"`
+
```
Comment (fatsheep9146) [inline processor/k8sattributesprocessor/config.go:42]: If add new fields to config, the `README.md` should also be updated.

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 5]
```diff
@@ -179,50 +184,77 @@ func New(logger *zap.Logger, apiCfg k8sconfig.APIConfig, rules ExtractionRules,
...
 		})
 		if err != nil {
-			c.logger.Error("error adding event handler to node informer", zap.Error(err))
+			return err
 		}
+		synced = append(synced, reg.HasSynced)
 		go c.nodeInformer.Run(c.stopCh)
 	}
+
+	ok := make(chan struct{})
```
Comment (fatsheep9146) [inline processor/k8sattributesprocessor/internal/kube/client.go:237]: The channel variable named "ok" is used for what purpose?

**Review by fatsheep9146 — COMMENTED** [review]
Thanks for your patience! 

I have an advice about the two new options `start_timeout` and `error_when_timeout`.  

The reason for this enhancement is to make sure the k8sattributes processor will start consuming the incoming traces or metrics until the initial synchronization are completed, 
otherwise, some traces/metrics/logs may be unassociated with k8s metadata. 

But I think some users may care more about the start rate of collector than the completeness of the k8s metadata. 

So how about we change this two fields to 
- `force_wait_for_cache_synced bool` default is false
- `wait_for_synced_timeout time.Duration`

The users who care more about the completeness could set `force_wait_for_cache_synced bool` to true and if the synced is still not completed after timeout `wait_for_synced_timeout`, the collector should directly return error instead of running, since the completeness could not be guaranteed.

@h0cheung @TylerHelmuth

**Comment by h0cheung** [comment]
> * force_wait_for_cache_synced

I agree. What are your thoughts?  @TylerHelmuth

**Comment by orloffv** [comment]
@TylerHelmuth can you see, plz?

**Comment by TylerHelmuth** [comment]
@fatsheep9146 I like that idea.  I suggest adjusting naming it `wait_for_metadata` and `wait_for_metadata_timeout` to hide some of the details from the end user.  I _believe_ we use the term `metadata` already to refer to the pod data that is gathered and held.

**Comment by h0cheung** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

remove stale

**Comment by h0cheung** [comment]
> @fatsheep9146 I like that idea. I suggest adjusting naming it `wait_for_metadata` and `wait_for_metadata_timeout` to hide some of the details from the end user. I _believe_ we use the term `metadata` already to refer to the pod data that is gathered and held.

Thanks. These two configuration items look simple and clear. I think we can use them.

**Comment by valner** [comment]
@h0cheung Hi, could you take some time to work out the code and merge the pull request? I’d really appreciate your update on this. Thanks a lot

File: processor/k8sattributesprocessor/config.go (hunk 4 above)
Comment (h0cheung) [reply processor/k8sattributesprocessor/config.go:50]: done

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 5 above)
Comment (h0cheung) [reply processor/k8sattributesprocessor/internal/kube/client.go]: It's used to run together with the waiting api of k8s. Now it has been refactored to be a timeoutCh, which will stop the `WaitForCacheSync` after timeout reached.

File: .chloggen/k8sattributes-block.yaml (hunk 3 above)
Comment (h0cheung) [reply .chloggen/k8sattributes-block.yaml]: done

File: processor/k8sattributesprocessor/processor.go (hunk 1 above)
Comment (h0cheung) [reply processor/k8sattributesprocessor/processor.go]: After I made all errors returned as is, the lifecycle unit tests fails with `unable to load k8s config, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined`. Should we make further changes to make it start normally with default configs and no environment variables?

**Comment by orloffv** [comment]
@fatsheep9146, @TylerHelmuth  please take a look

**Comment by fatsheep9146** [comment]
@h0cheung sorry for delay response，could you solve the conflict?

**Comment by valner** [comment]
@h0cheung
Hi there! I was wondering if you could possibly take a little more time to merge the feature? I’d really appreciate it!

**Comment by orloffv** [comment]
Hello @h0cheung,

We wanted to thank you for your work on this pull request. It is very important to us because the related bug significantly affects our work. If you need any assistance or information to speed up the process, please let us know. We appreciate your efforts and look forward to its completion.

Thank you!

**Comment by h0cheung** [comment]
Updated and skipped the lifecycle tests because of failing without KUBERNETES_SERVICE_HOST.

How does `prometheus-compliance-tests` work? Is its failure caused by this pr?

**Comment by valner** [comment]
It seems like `prometheus-compliance-tests` are flaky on the main as mentioned in the issue
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35119

**Comment by valner** [comment]
Seems like flaky tests were fixed as mentioned in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35119

Therefore, @h0cheung could you please rebase one more time?

**Comment by orloffv** [comment]
@TylerHelmuth pls take a look

**Comment by valner** [comment]
@h0cheung As I see in the tests we need to run `go mod tidy` to fix the build. Could you please do it when you have time?

**Comment by valner** [comment]
@h0cheung Hi!
It would be wonderful if you have time to do `go mod tidy` and push changes to the branch.

**Comment by orloffv** [comment]
@TylerHelmuth pls take a look

File: processor/k8sattributesprocessor/metadata.yaml [hunk 6]
```diff
@@ -113,6 +113,7 @@ resource_attributes:
 
 tests:
   config:
+  skip_lifecycle: true
```
Comment (TylerHelmuth) [inline processor/k8sattributesprocessor/metadata.yaml:116]: Why is this needed now?

File: processor/k8sattributesprocessor/internal/kube/client_test.go
Comment (TylerHelmuth) [inline processor/k8sattributesprocessor/internal/kube/client_test.go:1]: Can you add a test for `Start` that checks the timeout scenario and the error returned?

File: processor/k8sattributesprocessor/metadata.yaml (hunk 6 above)
Comment (h0cheung) [reply processor/k8sattributesprocessor/metadata.yaml:116]: After this change, any failure of the kube client will be returned to the main goroutine: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32622/files#diff-952fd1650b30a58e23e3895568409f67751fa822f3557cd2c6c6842a4700ee88L75.

Unfortunately, the kube client consistently fails to run without proper configuration or environment variables, returning the following error: unable to load k8s config, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined. This breaks the lifecycle test.

This error originates from internal/k8sconfig. As of now, I have not found an effective way to pass the lifecycle test.

File: processor/k8sattributesprocessor/internal/kube/client_test.go
Comment (h0cheung) [reply processor/k8sattributesprocessor/internal/kube/client_test.go:1]: done

File: processor/k8sattributesprocessor/metadata.yaml (hunk 6 above)
Comment (TylerHelmuth) [reply processor/k8sattributesprocessor/metadata.yaml:116]: Is there anything you can set in `tests.config` (thats the component's config for these tests) that would handle this?

Comment (h0cheung) [reply processor/k8sattributesprocessor/metadata.yaml:116]: No. Only `auth_type` is configurable, and all auth types depend on whether environment variables or config files in `~/.kube`.

Comment (TylerHelmuth) [reply processor/k8sattributesprocessor/metadata.yaml:116]: Seeing as how the k8scluster and k8sobjects recevier suffer from this as well it is ok.

**Review by TylerHelmuth — APPROVED** [review]

File: .chloggen/k8sattributes-block.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: processor/k8sattributes
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Block when starting util the metadata have been synced, to fix that some data couldn't be associated with metadata when the agent was just started.
```
Comment (ChrsMark) [inline .chloggen/k8sattributes-block.yaml:10]:
```suggestion
note: Block when starting until the metadata have been synced, to fix that some data couldn't be associated with metadata when the agent was just started.
```

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 8]
```diff
@@ -84,21 +87,23 @@ var rRegex = regexp.MustCompile(`^(.*)-[0-9a-zA-Z]+$`)
 var cronJobRegex = regexp.MustCompile(`^(.*)-[0-9]+$`)
 
 // New initializes a new k8s Client.
-func New(set component.TelemetrySettings, apiCfg k8sconfig.APIConfig, rules ExtractionRules, filters Filters, associations []Association, exclude Excludes, newClientSet APIClientsetProvider, newInformer InformerProvider, newNamespaceInformer InformerProviderNamespace, newReplicaSetInformer InformerProviderReplicaSet) (Client, error) {
+func New(set component.TelemetrySettings, apiCfg k8sconfig.APIConfig, rules ExtractionRules, filters Filters, associations []Association, exclude Excludes, newClientSet APIClientsetProvider, newInformer InformerProvider, newNamespaceInformer InformerProviderNamespace, newReplicaSetInformer InformerProviderReplicaSet, waitForMetadata bool, waitForMetadataTimeout time.Duration) (Client, error) {
```
Comment (ChrsMark) [inline processor/k8sattributesprocessor/internal/kube/client.go:90]: nit: since we are on it, could you split the declaration with one parameter per line to make the diffs easier?

File: processor/k8sattributesprocessor/README.md [hunk 9]
```diff
@@ -196,6 +196,21 @@ the processor associates the received trace to the pod, based on the connection
 }
 ```
 
+By default, the processor will be ready as soon as it starts, even no metadata has been fetched yet.
```
Comment (ChrsMark) [inline processor/k8sattributesprocessor/README.md:199]:
```suggestion
By default, the processor will be ready as soon as it starts, even if no metadata has been fetched yet.
```

File: processor/k8sattributesprocessor/README.md [hunk 10]
```diff
@@ -196,6 +196,21 @@ the processor associates the received trace to the pod, based on the connection
...
+If data is sent to this processor before the metadata is synced, there will be no metadata to enrich the data with.
+
+To wait for the metadata to be synced before the processor is ready, set the `wait_for_metadata` option to `true`.
+Then the processor will not be ready until the metadata has been synced.
```
Comment (ChrsMark) [inline processor/k8sattributesprocessor/README.md:203]: Does this mean that the Collector is blocked from getting started until the processor becomes ready?
If that's the case, maybe this should be mentioned explicitly here.

**Review by ChrsMark — APPROVED** [review]
LGTM

Left some nits.

File: processor/k8sattributesprocessor/README.md (hunk 10 above)
Comment (h0cheung) [reply processor/k8sattributesprocessor/README.md:203]:
```suggestion
Then the processor will not be ready until the metadata is fully synced. As a result, the start-up of the Collector tool will be blocked. If the metadata cannot be synced, the Collector tool will ultimately fail to start.
```

Comment (h0cheung) [reply processor/k8sattributesprocessor/README.md:203]: Yes. I'll add it.

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 8 above)
Comment (h0cheung) [reply processor/k8sattributesprocessor/internal/kube/client.go]: done

**Review by ChrsMark — APPROVED** [review]

File: processor/k8sattributesprocessor/README.md [hunk 11]
```diff
@@ -196,6 +196,21 @@ the processor associates the received trace to the pod, based on the connection
 }
 ```
 
+By default, the processor will be ready as soon as it starts, even if no metadata has been fetched yet.
+If data is sent to this processor before the metadata is synced, there will be no metadata to enrich the data with.
+
+To wait for the metadata to be synced before the processor is ready, set the `wait_for_metadata` option to `true`.
+Then the processor will not be ready until the metadata is fully synced. As a result, the start-up of the Collector tool will be blocked. If the metadata cannot be synced, the Collector tool will ultimately fail to start.
```
Comment (ChrsMark) [inline processor/k8sattributesprocessor/README.md:203]:
```suggestion
Then the processor will not be ready until the metadata is fully synced. As a result, the start-up of the Collector will be blocked. If the metadata cannot be synced, the Collector will ultimately fail to start.
```

**Review by ChrsMark — COMMENTED** [review]
One minor wording improvement :).

**Review by ChrsMark — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@fatsheep9146 @dmitryax I plan to merge by end of week, please comment if you have objections

**Review by fatsheep9146 — APPROVED** [review]
LGTM! Thanks
